client.go

```diff
@@ -949,17 +949,18 @@ func (c *Client) sniffNode(ctx context.Context, url string) []*conn {
 	c.mu.RUnlock()
 
 	res, err := c.c.Do((*http.Request)(req).WithContext(ctx))
+
+	//Make sure resp.Body is closed, if it exists, regardless of error state
+	if res != nil && res.Body != nil {
```

Comment:
Actually that's not necessary with `*http.Client.Do`. Its documentation clearly states:

> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

I'm not sure why it's a problem with the AWS signing transport, but I've never seen an issue with non-AWS clusters (and haven't seen any reports of such problems from others).

Reply:
Also, see [`http.Response`](https://golang.org/pkg/net/http/#Response):

```
// The http Client and Transport guarantee that Body is always
// non-nil, even on responses without a body or responses with
// a zero-length body. It is the caller's responsibility to
// close Body.
```

Reply:
You are mis-reading Client.Do()  It says if error is nil, than there is a body.
What it DIDN'T say, is if error is NOT nil, than there is no body to close.

Your network leak, can come if you get a 404 error, with a non-nil body, because original code aborted on error (before closing Body)

Reply:
The `net/http` gives an explicit example of how to close the response body in its Overview section. The code is:

```
resp, err := http.Get("http://example.com/")
if err != nil {
	// handle error
}
defer resp.Body.Close()
```

Reply:
Below that `If the returned error is nil ...` section, documentation goes on with:

> On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed.

... which Brad Fitzpatrick [confirms here](https://github.com/golang/go/issues/17780#issuecomment-258293699).